File: 2721-sum-of-distances/sum-of-distances.java
class Solution {
    public long[] distance(int[] nums) {
         int n = nums.length;
        long[] ans = new long[n];

        HashMap<Integer, List<Integer>> map = new HashMap<>();

        // Group indices by value
        for(int i = 0; i < n; i++){
            map.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }

        for(List<Integer> pos : map.values()){

            int m = pos.size();
            long[] prefix = new long[m];

            prefix[0] = pos.get(0);

            for(int i=1; i<m; i++){
                prefix[i] = prefix[i-1] + pos.get(i);
            }

            for(int i=0; i<m; i++){

                long idx = pos.get(i);

                long left = (i==0) ? 0 :
                        (long)i * idx - prefix[i-1];

                long right = (i==m-1) ? 0 :
                        (prefix[m-1]-prefix[i])
                        - (long)(m-i-1)*idx;

                ans[(int)idx] = left + right;
            }
        }

        return ans;
    }
}

File: 2721-sum-of-distances/sum-of-distances.py
class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        groups = defaultdict(list)

        # Store indices for each value
        for i, num in enumerate(nums):
            groups[num].append(i)

        ans = [0] * len(nums)

        for pos in groups.values():
            m = len(pos)

            # Prefix sum of indices
            prefix = [0] * m
            prefix[0] = pos[0]

            for i in range(1, m):
                prefix[i] = prefix[i - 1] + pos[i]

            for i, idx in enumerate(pos):

                left = 0 if i == 0 else i * idx - prefix[i - 1]

                right = 0 if i == m - 1 else \
                    (prefix[-1] - prefix[i]) - (m - i - 1) * idx

                ans[idx] = left + right

        return ans